Add value and info APIs (#492)
pmix/pmix-standard: Chap_API_Struct.tex

\argin{n}{Number of structures in the array (\code{size_t})}
\end{arglist}

%%%%%%%%%%%
\littleheader{Get size of data in a value structure}
\declareapiProvisional{PMIx_Value_get_size}

%%%%
\summary

Compute and return the size (in bytes) of the data payload in a \refstruct{pmix_value_t} structure.

%%%%
\format

\versionMarker{5.1}
\cspecificstart
\begin{codepar}
pmix_status_t
PMIx_Value_get_size(const pmix_value_t *val,
                    size_t *size);
\end{codepar}
\cspecificend

\begin{arglist}
\argin{val}{The \refstruct{pmix_value_t} where the data is located (pointer to \refstruct{pmix_value_t})}

> \argin{size}{Pointer to variable where size will be returned}

Author: ```suggestion
\argout{size}{Pointer to variable where size will be returned}
```

\argin{n}{Number of structures in the array (\code{size_t})}
\end{arglist}

%%%%%%%%%%%
\littleheader{Get size of data in an info structure}
\declareapiProvisional{PMIx_Info_get_size}

%%%%
\summary

Compute and return the size (in bytes) of the data payload in an \refstruct{pmix_info_t} structure.

%%%%
\format

\versionMarker{5.1}
\cspecificstart
\begin{codepar}
pmix_status_t
PMIx_Info_get_size(const pmix_info_t *info,
                   size_t *size);
\end{codepar}
\cspecificend

\begin{arglist}
\argin{info}{The \refstruct{pmix_info_t} where the data is located}

> \argin{type}{Pointer to variable where size will be returned}

Author: ```suggestion
\argout{size}{Pointer to variable where size will be returned}
```

Copy the provided key and data into a \refstruct{pmix_info_t} on the list. The key and any data stored in the source value can be modified or free'd without affecting the copied data once the function has completed.

%%%%%%%%%%%
\littleheader{Retrieve next \refstruct{pmix_info_t} on a list}
\declareapiProvisional{PMIx_Info_list_get_info}

%%%%
\summary

Retrieve next \refstruct{pmix_info_t} on a list.

%%%%
\format

> \versionMarker{5.0}

Author: ```suggestion
\versionMarker{5.1}
```

\cspecificstart
\begin{codepar}
pmix_info_t *

> PMIx_Info_list_get_info(void *ptr,

Author: ```suggestion
PMIx_Info_list_get_info(void *head,
```

Author: While I think this is a good suggestion, the rest of the `PMIx_Info_list_*` APIs use the `ptr` name for the list being passed in. If we decide to rename the parameter, we should do all the list APIs at once, IMO.

Reviewer: Agreed

>                         void *prev,

Author: ```suggestion
                        void *curr,
```

                        void **next);
\end{codepar}
\cspecificend

\begin{arglist}
\argin{ptr}{A \code{void*} pointer initialized via \refapi{PMIx_Info_list_start} (handle)}

> \argin{prev}{Pointer to previous list element}

Author: ```suggestion
\argin{curr}{Pointer to current list element}
```

\argout{next}{Pointer to next list element}
\end{arglist}

> Returns \refconst{PMIX_SUCCESS} or a negative value corresponding to a PMIx error constant.

Author: figure out if return type should be `pmix_status_t`. If not, update this description.

Author: Return value should never be null. It is the contents of the `prev` object.

> Returns \refconst{PMIX_SUCCESS} or a negative value corresponding to a PMIx error constant.

Author: ```suggestion
```

%%%%
\descr

> Retrieve the next info on the list. Passing NULL to the prev parameter will return the first pmix_info_t on the list. If next is set to NULL, it indicates the end of the list.

Author: Update this with new arg names prev->curr. "Retrieves current info and sets next pointer"

> Retrieve the next info on the list. Passing NULL to the prev parameter will return the first pmix_info_t on the list. If next is set to NULL, it indicates the end of the list.

Author: ```suggestion
Retrieve the pmix_info_t object pointed to by curr from a list. Passing NULL to the curr parameter will return the first pmix_info_t on the list. If next is set to NULL, it indicates the end of the list.
```

%%%%%%%%%%%
\littleheader{Prepend a \refstruct{pmix_info_t} structure to a list}
\declareapiProvisional{PMIx_Info_list_prepend}

%%%%
\summary

Prepend a \refstruct{pmix_info_t} structure containing the specified value to the provided list.

%%%%
\format

\versionMarker{5.1}
\cspecificstart
\begin{codepar}
pmix_status_t

> PMIx_Info_list_prepend(void *ptr,

Author: ```suggestion
PMIx_Info_list_prepend(void *head,
```

                       const char *key,
                       const void *value,
                       pmix_data_type_t type);
\end{codepar}
\cspecificend

\begin{arglist}

> \argin{ptr}{A \code{void*} pointer initialized via \refapi{PMIx_Info_list_start} (handle)}

Author: ```suggestion
\argin{head}{A \code{void*} pointer initialized via \refapi{PMIx_Info_list_start} (handle)}
```

> \argin{key}{String key to be loaded - must be less than or equal to \refconst{PMIX_MAX_KEYLEN} in length (handle)}

Author: ```suggestion
\argin{key}{String key to be loaded - must be less than or equal to \refconst{PMIX_MAX_KEYLEN} in length}
```

> \argin{value}{Pointer to the data value to be loaded (handle)}

Author: ```suggestion
\argin{value}{Pointer to the data value to be loaded}
```
